vfs_handlers/s3_handlers.cc

```diff
@@ -425,306 +549,266 @@ makeStreamingDownload(const std::string & bucket,
-        std::tie(impl->bucket, impl->object) = S3Api::parseUri(urlStr);
-        impl->metadata = metadata;
-        impl->onException = excCallback;
-        impl->chunkSize = 8 * 1024 * 1024;  // start with 8MB and ramp up
+/****************************************************************************/
+/* S3 UPLOADER                                                              */
+/****************************************************************************/
 
-        impl->start();
-    }
+inline void touchByte(const char * c)
+{
+    __asm__(" # [in]":: [in] "r" (*c):);
+}
 
-    typedef char char_type;
-    struct category
-        : public boost::iostreams::output,
-          public boost::iostreams::device_tag,
-          public boost::iostreams::closable_tag
-    {
-    };
+inline void touch(const char * start, size_t size)
+{
+    const char * current = start - (intptr_t) start % 4096;
+    if (current < start) {
+        current += 4096;
+    }
+    const char * end = start + size;
+    for (; current < end; current += 4096) {
+        touchByte(current);
+    }
+}
 
-    struct Impl {
-        Impl()
-            : offset(0), chunkIndex(0), shutdown(false),
-              chunks(16)
-        {
-        }
 
-        ~Impl()
-        {
-            //cerr << "destroying streaming upload at " << object << endl;
-            stop();
+struct S3Uploader {
+    S3Uploader(const S3Api * api,
+               const string & bucket,
+               const string & resource, // starts with "/", unescaped (buggy)
+               const OnUriHandlerException & excCallback,
+               const S3Api::ObjectMetadata & objectMetadata)
+        : api(api),
+          bucket(bucket), resource(resource),
+          metadata(objectMetadata),
+          onException(excCallback),
+          closed(false),
+          chunkSize(8 * 1024 * 1024), // start with 8MB and ramp up
+          currentRq(0),
+          activeRqs(0)
+    {
+        /* Maximum chunk size is what we can do in 3 seconds, up to 1% of
+           system memory. */
+#if 0
+        maxChunkSize = api->bandwidthToServiceMbps * 3.0 * 1000000;
+        size_t sysMemory = getTotalSystemMemory();
+        maxChunkSize = std::min(maxChunkSize, sysMemory / 100);
+#else
+        maxChunkSize = 64 * 1024 * 1024;
+#endif
+
+        try {
+            S3Api::MultiPartUpload upload
+              = api->obtainMultiPartUpload(bucket, resource, metadata,
+                                           S3Api::UR_EXCLUSIVE);
+            uploadId = upload.id;
         }
-
-        shared_ptr<S3Api> owner;
-        std::string bucket;
-        std::string object;
-        S3Api::ObjectMetadata metadata;
-        std::string uploadId;
-        size_t offset;
-        size_t chunkSize;
-        size_t chunkIndex;
-        atomic<bool> shutdown;
-        std::vector<std::thread> tg;
-
-        Date startDate;
-
-        struct Chunk {
-            Chunk() : data(nullptr)
-            {
+        catch (...) {
```

Comment:
We wouldn't want the onException function to run on thread cancellation, so I'd say in this case yes, even if rethrowing.